python/pyspark/sql/connect/client/artifact.py

```diff
@@ -0,0 +1,270 @@
+            nonlocal current_batch
+            nonlocal current_batch_size
+
+            yield from self._add_batched_artifacts(current_batch)
+            current_batch = []
+            current_batch_size = 0
+
+        for artifact in artifacts:
+            data = artifact.storage
+            size = data.size
+            if size > ArtifactManager.CHUNK_SIZE:
+                # Payload can either be a batch OR a single chunked artifact.
+                # Write batch if non-empty before chunking current artifact.
+                if len(current_batch) > 0:
+                    yield from write_batch()
+                yield from self._add_chunked_artifact(artifact)
+            else:
+                if current_batch_size + size > ArtifactManager.CHUNK_SIZE:
+                    yield from write_batch()
+                add_to_batch(artifact, size)
+
+        if len(current_batch) > 0:
+            yield from write_batch()
+
+    def _add_batched_artifacts(
+        self, artifacts: Iterable[Artifact]
+    ) -> Iterator[proto.AddArtifactsRequest]:
+        """
+        Add a batch of artifacts to the stream. All the artifacts in this call are packaged into a
+        single :class:`proto.AddArtifactsRequest`.
+        """
+        artifact_chunks = []
+
+        for artifact in artifacts:
+            binary = artifact.storage.stream.read()
+            crc32 = zlib.crc32(binary)
+            data = proto.AddArtifactsRequest.ArtifactChunk(data=binary, crc=crc32)
+            artifact_chunks.append(
+                proto.AddArtifactsRequest.SingleChunkArtifact(name=artifact.path, data=data)
+            )
+
+        # Write the request once
+        yield proto.AddArtifactsRequest(
+            session_id=self._session_id,
+            user_context=self._user_context,
+            batch=proto.AddArtifactsRequest.Batch(artifacts=artifact_chunks),
+        )
+
+    def _add_chunked_artifact(self, artifact: Artifact) -> Iterator[proto.AddArtifactsRequest]:
+        """
+        Add a artifact in chunks to the stream. The artifact's data is spread out over multiple
+        :class:`proto.AddArtifactsRequest requests`.
+        """
+        initial_batch = True
+        # Integer division that rounds up to the nearest whole number.
+        get_num_chunks = int(
+            (artifact.size + (ArtifactManager.CHUNK_SIZE - 1)) / ArtifactManager.CHUNK_SIZE
+        )
+
+        # Consume stream in chunks until there is no data left to read.
+        for chunk in iter(lambda: artifact.storage.stream.read(ArtifactManager.CHUNK_SIZE), b""):
+            if initial_batch:
+                # First RPC contains the `BeginChunkedArtifact` payload (`begin_chunk`).
+                yield proto.AddArtifactsRequest(
+                    session_id=self._session_id,
+                    user_context=self._user_context,
+                    begin_chunk=proto.AddArtifactsRequest.BeginChunkedArtifact(
```

Comment:
I feel like a separate design schema is we have `open`, `transfer`, `close` RPC API to streaming chunks.

Just a thought and we can use existing design as it is.